Mathlib/Combinatorics/Schnirelmann.lean

```diff
@@ -0,0 +1,247 @@
+which reduces the proof obligations later that would arise with `Nat.card`.
+
+## TODO
+* Give other calculations of the density, for example powers and their sumsets.
+* Define other densities like the lower and upper asymptotic density, and the natural density,
+  and show how these relate to the Schnirelmann density.
+* Show that if the sum of two densities is at least one, the sumset covers the positive naturals.
+* Prove Schnirelmann's theorem and Mann's theorem on the subadditivity of this density.
+
+## References
+Ruzsa, Imre. (2009). Sumsets and structure. Combinatorial Number Theory and Additive Group Theory.
+-/
+
+open Finset
+
+/-- The Schnirelmann density is defined as the infimum of |A ∩ {1, ..., n}| / n as n ranges over
+the positive naturals. -/
+noncomputable def schnirelmannDensity (A : Set ℕ) [DecidablePred (· ∈ A)] : ℝ :=
+  ⨅ n : {n : ℕ // 0 < n}, ((Ioc (0 : ℕ) n).filter (· ∈ A)).card / n
+
+section
+
+variable {A : Set ℕ} [DecidablePred (· ∈ A)]
+
+lemma schnirelmannDensity_nonneg : 0 ≤ schnirelmannDensity A :=
+  Real.iInf_nonneg (fun _ => by positivity)
+
+lemma schnirelmannDensity_le_div {n : ℕ} (hn : n ≠ 0) :
+    schnirelmannDensity A ≤ ((Ioc 0 n).filter (· ∈ A)).card / n :=
+  ciInf_le ⟨0, fun _ ⟨_, hx⟩ => hx ▸ by positivity⟩ (⟨n, hn.bot_lt⟩ : {n : ℕ // 0 < n})
+
+/--
+For any natural `n`, the Schnirelmann density multiplied by `n` is bounded by `|A ∩ {1, ..., n}|`.
+Note this property fails for the natural density.
+-/
+lemma schnirelmannDensity_mul_le_card_filter {n : ℕ} :
+    schnirelmannDensity A * n ≤ ((Ioc 0 n).filter (· ∈ A)).card := by
+  rcases eq_or_ne n 0 with rfl | hn
+  · simp
+  exact (le_div_iff (by positivity)).1 (schnirelmannDensity_le_div hn)
+
+/--
+To show the Schnirelmann density is upper bounded by `x`, it suffices to show
+`|A ∩ {1, ..., n}| / n ≤ x`, for any chosen positive value of `n`.
+
+We provide `n` explicitly here to make this lemma more easily usable in `apply` or `refine`.
+This lemma is analogous to `ciInf_le_of_le`.
+-/
+lemma schnirelmannDensity_le_of_le {x : ℝ} (n : ℕ) (hn : n ≠ 0)
+    (hx : ((Ioc 0 n).filter (· ∈ A)).card / n ≤ x) : schnirelmannDensity A ≤ x :=
+  (schnirelmannDensity_le_div hn).trans hx
+
+lemma schnirelmannDensity_le_one : schnirelmannDensity A ≤ 1 :=
+  schnirelmannDensity_le_of_le 1 one_ne_zero <|
+    by rw [Nat.cast_one, div_one, Nat.cast_le_one]; exact card_filter_le _ _
+
+/--
+If `k` is omitted from the set, its Schnirelmann density is upper bounded by `1 - 1 / k`.
+-/
+lemma schnirelmannDensity_le_of_not_mem {k : ℕ} (hk : k ∉ A) :
+    schnirelmannDensity A ≤ 1 - 1 / k := by
+  rcases k.eq_zero_or_pos with rfl | hk'
+  · simpa using schnirelmannDensity_le_one
+  apply schnirelmannDensity_le_of_le k hk'.ne'
+  rw [one_sub_div (Nat.cast_pos.2 hk').ne']
+  apply div_le_div_of_le (Nat.cast_nonneg _)
+  rw [←Nat.cast_pred hk', Nat.cast_le]
+  suffices : (Ioc 0 k).filter (· ∈ A) ⊆ Ioo 0 k; exact (card_le_of_subset this).trans_eq (by simp)
```

Comment:
```suggestion
  gcongr
  rw [←Nat.cast_pred hk', Nat.cast_le]
  calc _ ≤ card (Ioo 0 k) := ?_
    _ = _ := by simp
  gcongr
```

Reply:
I originally tried many of these already, but they made the file slower to compile for me, and makes the proofs significantly more brittle, so I'm very hesitant to put them in (just like nonterminal `simp`s)!

Reply:
Interesting, we could try to get data here ... in the mean time the equivalent of `simp only` would be to use patterns, `gcongr ?_ / _` and `gcongr card ?_`.  More worthwhile when you haven't already looked up the lemmas though.

Reply:
In particular, I intended `gcongr` to be a mid-proof (rather than finishing) tactic more often than not, so if there is evidence of a slowdown, we have work to do.  I'm less worried about brittleness -- such changes ought to be easy to repair.

Reply:
Ah the use of patterns would be nice, although as you say I already knew where these lemmas were! 
I'm not so sure they're easy to repair - particularly in the case where a congruence lemma is applied which goes too far, and it can be hard to figure out which step was unhelpful when repairing. Using patterns would help with this though!